[FLOC-3372] Vagrant tutorial box fix for Docker 1.9.0

sarum90 opened the pull request:
It appears that docker 1.9.0 broke the Vagrant 'docker' provisioner on our tutorial's CentOS 7 box.

Currently the buildbot run for `flocker/vagrant/build/tutorial` has been red with failures that always end in:
```
==> default: Running provisioner: docker...
    default: Configuring Docker to autostart containers...
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

sed -i.bak 's/docker -d/docker -d -r=true/' /etc/sysconfig/docker 

Stdout from the command:



Stderr from the command:

sed: can't read /etc/sysconfig/docker: No such file or directory
Failed 1: vagrant up
```
This seems to indicate that the Vagrant 'docker' provisioner assumes that docker will have `/etc/sysconfig/docker`, which is evidently no longer true with docker 1.9.0 on CentOS 7.  Since we were only using the docker provisioner for the sake of starting docker, and pre-caching docker images, we can just do those things in a shell provisioner rather than relying on the Vagrant 'docker' provisioner.

Interestingly this is effectively just a flip-flop of https://github.com/ClusterHQ/flocker/commit/c4e34185327a101cc309055e4094615691fdc03d, although I cannot find any documentation as to why we switched to using the 'docker' provider. Hopefully we only started using it because it was so tempting and not because the shell script idea is fundamentally flawed.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/clusterhq/flocker/2128)
<!-- Reviewable:end -->

exarkun commented on vagrant/tutorial/setup_docker.sh:
```diff
@@ -0,0 +1,26 @@
+IMAGES+=" clusterhq/mongodb"
+IMAGES+=" redis"
+IMAGES+=" python:2.7-slim"
+IMAGES+=" clusterhq/flask"
```
IMAGES=(busybox clusterhq/mongodb redis python:2.7-slim clusterhq/flask)

sarum90 replied:
Done.

sarum90 replied:
Using quotes instead.

sarum90 replied:
Decided to grep through the bash man page to figure out what was going on. We could have done this, but then below we would need ${IMAGES[*]} ( or "${IMAGES[@]}" might be even better). I'll keep that in mind for next time, although honestly the fact that ${IMAGES} just returns ${IMAGES[0]} makes me sort of not want to use bash Arrays.

For now leaving as quotes given that I've already restarted the tests.

exarkun commented on vagrant/tutorial/setup_docker.sh:
```diff
@@ -0,0 +1,26 @@
+IMAGES+=" redis"
+IMAGES+=" python:2.7-slim"
+IMAGES+=" clusterhq/flask"
+for I in $IMAGES
```
${IMAGES}

sarum90 replied:
Done.

exarkun commented on vagrant/tutorial/setup_docker.sh:
```diff
@@ -0,0 +1,26 @@
+IMAGES+=" clusterhq/flask"
+for I in $IMAGES
+do
+    echo Pulling image $I...
```
${I}

sarum90 replied:
Done.

exarkun commented on vagrant/tutorial/setup_docker.sh:
```diff
@@ -0,0 +1,26 @@
+for I in $IMAGES
+do
+    echo Pulling image $I...
+    docker pull $I
```
${I}

sarum90 replied:
Done.

exarkun commented:
c4e34185327a101cc309055e4094615691fdc03d seems noteworthy (or the set up for the joke to which this is the punchline, not sure which)

exarkun commented on vagrant/tutorial/Vagrantfile:
```diff
@@ -29,13 +29,9 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
-    'python:2.7-slim',
-    'clusterhq/flask'
-  ]
+  config.vm.provision :shell,
```
Perhaps worth some commentary that the Vagrant Docker provisioner is fragile and we're intentionally not using it even though it's obviously the correct thing to use.

sarum90 replied:
Added.

exarkun commented:
Thanks.  Please twiddle the shell script as suggested and add some warnings to try to keep us from flip-flopping like this /again/... then merge.

sarum90 commented:
Addressed comments (including adding referrence to the flip-flop to the pull request description), going to re-run the tests at least until the point where the shell script is actually run, then merge.

sarum90 commented:
Not submitting: Looking at the logs: http://build.clusterhq.com/builders/flocker%2Fvagrant%2Fbuild%2Ftutorial/builds/2292/steps/build-base-box/logs/stdio

We are not pulling all of the images now. Doing some local hacking to figure out why.

sarum90 commented:
```
IMAGES=(busybox clusterhq/mongodb redis python:2.7-slim clusterhq/flask)
for I in ${IMAGES}
do
    echo HI ${I}
done
```
yields `HI busybox`

whereas:
```
IMAGES="busybox clusterhq/mongodb redis python:2.7-slim clusterhq/flask"
for I in ${IMAGES}
do
    echo HI ${I}
done
```
yields:
```
HI busybox
HI clusterhq/mongodb
HI redis
HI python:2.7-slim
HI clusterhq/flask
```

I'm going to use quotes instead of parens.

sarum90 commented:
Okay, latest run looks better: http://build.clusterhq.com/builders/flocker%2Fvagrant%2Fbuild%2Ftutorial/builds/2293/steps/build-base-box/logs/stdio

I'll be merging shortly.

exarkun commented:
> yields HI busybox

Oops, sorry.  Too late o'clock for writing shell.  I gave you the wrong expansion expression.  `X=(...)` defines an array.  To expand the full array you need `${X[*]}` instead of `${X}`.

Not a big deal.  Just posting for the sake of completeness.  Also, to have a chance to point out that shell sucks.
